### ISSUE TITLE
globally desaturated colors on Canon M6 Mark II in DT 4.0.0 versus 3.8.1 (and probably Canon 90D in both)

### ISSUE THREAD

**Issue by douzzer** [darktable-org/darktable#12171]
I've just upgraded from 3.8.1 to 4.0.0, and am finding that color is desaturated on all images, whether already in the database, or freshly imported, and whether viewed inside the application, or exported to sRGB JPEGs.  Disabling/enabling filmic, exposure, and highlight reconstruction doesn't cause or cure the problem.

All the images are from a Canon M6 ii.

As an experiment in 4.0.0, I was able to roughly match correct saturation with a +32.5% chroma boost, but the colors still weren't quite right.

In all of the below image pairs, the export from 3.8.1 is first, 4.0.0 is second, with unchanged darktable image settings.

Quantitatively, the saturation percentages of the bottom row of color patches in the DT3.8.1 export are 63, 100, 100, 100, 83, 100, respectively, versus DT4.0.0 export 60, 100, 56, 86, 72, 73.  Viewing the physical color patches under reference illumination, it appeared that the saturation of the green patch deviated the most, consistent with the quantitative results.

![IMG_1768](https://user-images.githubusercontent.com/5521927/178851466-cdf10dfb-a4d9-496a-8f9e-2b04693e9d19.jpg)
![IMG_1768](https://user-images.githubusercontent.com/5521927/178851509-4636dee7-3888-479a-bad1-d70f38e2457e.jpg)

![IMG_1886](https://user-images.githubusercontent.com/5521927/178851663-d91ccc8e-d377-48da-9a56-66c78d05b628.jpg)
![IMG_1886](https://user-images.githubusercontent.com/5521927/178851687-fbba4ad4-6dc6-46b0-b0e6-f284b75c3f2a.jpg)

![IMG_0898](https://user-images.githubusercontent.com/5521927/178851525-6e5ecc33-c207-4864-af80-8b9bc64c6956.jpg)
![IMG_0898](https://user-images.githubusercontent.com/5521927/178851562-28f96b32-a084-4044-bcaf-31e08cca1237.jpg)

![IMG_1313](https://user-images.githubusercontent.com/5521927/178851589-fb1dde32-98ec-401c-991d-97feb46043d7.jpg)
![IMG_1313](https://user-images.githubusercontent.com/5521927/178851605-8a9f7372-50c9-462d-bc7d-748c1884b6ec.jpg)

![IMG_2584](https://user-images.githubusercontent.com/5521927/178851728-98135011-dcde-4d30-947f-a2d3da8f19ec.jpg)
![IMG_2584](https://user-images.githubusercontent.com/5521927/178851740-147a989a-5f6e-485a-b071-d97d51b4b822.jpg)

Both versions were built and run locally with identical build settings on Gentoo Linux.

**Comment by douzzer** [comment]
I've determined this behavior was caused by 845b003dd2f05a2878ce530a0f9ed0587e0bc3e9 , and affects only the Canon M6 Mark II.  I don't yet know for certain whether the new or old behavior is closer to the mark, but if there's a problem it would be in libraw, not in DT.

The improvements in DT4 itself are wonderful progress!

**Comment by douzzer** [comment]
Addendum: DT bundles libraw at `src/external/LibRaw`, so if libraw's color mappings for the M6-ii are wrong, DT will need to follow any resulting fixes.  It's a work in progress to evaluate the libraw transform's accuracy, but I think there is indeed an undersaturation problem.  libraw labeled the M6-ii (and identical 90D) primaries `/* temp */` at one point, then removed the `/* temp */` without changing the data, so it's presumptively suspect.  The R7 primaries may be higher quality and I'll be trying that next.

**Comment by douzzer** [comment]
The R7 primaries from libraw master are a better match for the M6-ii (and implicitly, the 90D) than the M6-ii "temp" data.

![IMG_1768](https://user-images.githubusercontent.com/5521927/179139566-18b4d757-601a-4cab-9234-363857119dd4.jpg)

A 3.5% global chroma boost is still needed to subjectively match the color references.  Notably, the resulting saturations are mostly the same as those obtained with the original M6 data.  The only deviation greater than 1.5% is magenta, 78.71% with R7 primaries vs 81.30% with M6 primaries.

But there are still lingering problems with greens and skin tones. The green patch, specifically, is so different between the M6 and M6-ii/R7 primaries that the former is out of sRGB gamut while the latter are not.  And on closer inspection, the hue of green differs by 9°, the most of any patch, with magenta a close second at 7°.

I strongly suspect the M6 primaries are closest to the mark, because they alone reproduce the slight pinkish look to the "blue" patches, readily apparent in the original reference.

So in the end I've opted to patch in the M6 primaries for the M6-ii, which gives distinctly more natural results, gives neutral saturation without fiddling, and -- not least -- doesn't require me to regrade my library.

I've reopened the ticket because, inevitably, other M6-ii + DT users will encounter this problem, and it probably makes sense to get ahead of it.

**Comment by douzzer** [comment]
I've updated the issue title to reflect that the Canon 90D uses the same pathologically desaturating primary matrix as the M6-ii, both in 3.8.1 and 4.0.0 (and upstream LibRaw).  The thread [here](https://discuss.pixls.us/t/canon-90d-cr3-very-poor-colour-rendering/23531) relates.  In that thread, [Glenn Butcher](https://discuss.pixls.us/u/ggbutcher) posts a matrix he derived from DPReview 90D images of the ColorChecker.  That matrix results in HSV color coordinates almost identical to those from the M6-i matrix, except for the saturated "blue" patch, where the saturation rises from 81.4% to 88.8%.  My initial impression is that this matrix gives the best results of any I've tried so far.

Glenn Butcher's matrix:
```
[ 10832, -1547, -1009 ],
[-3663, 9330, 2845 ],
[ -1112, 2234, 8395 ]
```

Results:

![IMG_1768_01](https://user-images.githubusercontent.com/5521927/179314507-f3538711-9fe6-4019-9cba-5fcda6cb7cc0.jpg)

**Comment by douzzer** [comment]
Followup on the above: in the end I'm using the R7 matrix for the M6-ii.  It seems quite high quality and, alone of the 3, produces passable results even without white balancing.  A knock-on benefit of that is that the DT-reported illumination color temp and tint in the WB module are bang-on, circa 5000K and 1.000 on a scene thusly illuminated, without any further calibration.

However, none of these matrices (M6-i, M6-ii/90D, R7, Glenn Butcher's) are usable (aside from casual stuff) without calibration.  I've now done that -- ColorChecker with fresh L/a/b values from an i1pro2 and ArgyllCMS patched into src/common/colorchecker.h -- and have found that it is apparently the R7 matrix that is closest to the mark, in terms of HSV values on the Q-13.  Btw, the in-the-release values in `xrite_24_2014` are quite close to what I measured, and I couldn't see an effect when I recomputed the correction matrices for my own measured values.

Here, for the record, is what the Q-13 scene actually looks like, with calibrated colors.  DT reports white temp is 5180K, tint is .991.

![IMG_1768_02](https://user-images.githubusercontent.com/5521927/180577962-d6a78dd3-cb5f-4458-a9c0-e52db1070fbd.jpg)

And here's what that tiger looks like with properly calibrated color:

![IMG_1886](https://user-images.githubusercontent.com/5521927/180578411-2a971111-1ec8-480a-98da-acce0d262dc7.jpg)

The M6-i matrix results, which DT users saw prior to the 4.0.0 release, are remarkably close!
